modules/core/src/main/java/org/apache/ignite/internal/processors/service/ServicesDeploymentTask.java

```diff
@@ -0,0 +1,868 @@
+     */
+    protected void onNodeLeft(UUID nodeId) {
+        initTaskFut.listen((IgniteInClosure<IgniteInternalFuture<?>>)fut -> {
+            if (isCompleted())
+                return;
+
+            final boolean crdChanged = nodeId.equals(crdId);
+
+            if (crdChanged) {
+                ClusterNode crd = srvcProc.coordinator();
+
+                if (crd != null) {
+                    crdId = crd.id();
+
+                    if (crd.isLocal())
+                        initCoordinator(evtTopVer);
+
+                    createAndSendSingleDeploymentsMessage(depId, depErrors);
+                }
+                else
+                    onAllServersLeft();
+            }
+            else if (ctx.localNodeId().equals(crdId)) {
+                synchronized (initCrdMux) {
+                    boolean rmvd = remaining.remove(nodeId);
+
+                    if (rmvd && remaining.isEmpty()) {
+                        singleDepsMsgs.remove(nodeId);
+
+                        onAllReceived();
+                    }
+                }
+            }
+        });
+    }
+
+    /**
+     * Handles case when all server nodes have left the grid.
+     */
+    private void onAllServersLeft() {
+        assert ctx.clientNode();
+
+        completeError(new ClusterTopologyServerNotFoundException("Failed to resolve coordinator to continue services " +
+            "deployment process: [locId=" + ctx.localNodeId() + "client=" + ctx.clientNode() + "evt=" + evt + ']'));
+    }
+
+    /**
+     * @return Cause discovery event.
+     */
+    public DiscoveryEvent event() {
+        return evt;
+    }
+
+    /**
+     * Returns cause of deployment process topology version.
+     *
+     * @return Cause of deployment process topology version.
+     */
+    public AffinityTopologyVersion topologyVersion() {
+        return evtTopVer;
+    }
+
+    /**
+     * Returns services deployment process id of the task.
+     *
+     * @return Services deployment process id.
+     */
+    public ServicesDeploymentProcessId deploymentId() {
+        return depId;
+    }
+
+    /**
+     * Completes the task.
+     */
+    public void completeSuccess() {
+        if (!completeStateFut.isDone()) {
+            completeInitiatingFuture(null);
+
+            completeStateFut.onDone();
+        }
+
+        if (!initTaskFut.isDone())
+            initTaskFut.onDone();
+
+        if (!initCrdFut.isDone())
+            initCrdFut.onDone();
+    }
+
+    /**
+     * @param err Error to complete with.
+     */
+    public void completeError(Throwable err) {
+        if (!completeStateFut.isDone()) {
+            completeInitiatingFuture(err);
+
+            completeStateFut.onDone(err);
+        }
+
+        if (!initTaskFut.isDone())
+            initTaskFut.onDone(err);
+
+        if (!initCrdFut.isDone())
+            initCrdFut.onDone(err);
+    }
+
+    /**
+     * Releases resources to reduce memory usages.
+     */
+    protected void clear() {
```

Comment:
The main reason is to release objects for GC as soon as possible.

Reply:
But you call it right after removing the task from `ServicesDeploymentManager#tasks` map. The entire task is ready for GC after that.

Reply:
`singleDepsMsgs` may be cached in communications spi;
Link on an entry of `depErrors` may be cached in `ServicesSingleDeploymentsMessage` also.

This depends on an algorithm of GC very much. For example, G1 splits memory on equal chunks and clean them separately, but objects can be placed in different chunks, e.g `singleDepsMsgs` in chunk `A` and its entry in `B`, and an object can't be cleaned from `B` while `singleDepsMsgs` from `A` is not cleaned.

Reply:
`DiscoveryCustomEvent#nullifyingCustomMsgLock` is used to avoid nullified messages which may be nullified by other systems.

For example, we listen to the following messages:
- `ChangeGlobalStateMessage` - to perform activation/deactivation actions; 
- `DynamicCacheChangeBatch` - to handle caches stopping to undeploy 
related affinity services; 
- `CacheAffinityChangeMessage `- to recalculate assignments for affinity 
services in case of late affinity. 

These messages may be nullified in PME infrastructure (`GridDhtPartitionsExchangeFuture#onDone`)

I've already faced with such case:
http://apache-ignite-developers.2346864.n4.nabble.com/Danger-change-of-DiscoveryCustomEvent-in-GridDhtPartitionsExchangeFuture-onDone-td35946.html